### PR TITLE
Don't follow non-local redirects from HTTP probes

### LINKED ISSUE #73172: kubelet's httpProbe  follow redirects
**What happened**:

httpProbe follows redirect

**What you expected to happen**:

301/302 code makes check successfull as per documentation

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:04:45Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.5", GitCommit:"753b2dbc622f5cc417845f0ff8a77f539a4213ea", GitTreeState:"clean", BuildDate:"2018-11-26T14:31:35Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

/sig node

### PR DESCRIPTION
**Pull request by tallclair**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

HTTP probes currently follow redirects, and perform the health check against the endpoint redirect to. This behavior differs from the documented behavior, which states that any HTTP response with a status code between 200 and 400 should be treated as success.

As a backwards-compatibility compromise, this PR modifies the probe logic follow redirects if they go to the same hostname (e.g. a relative redirect to a different path). If the redirect points to a different hostname, then the response is simply treated as a success.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/issues/73172

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
ACTION REQUIRED: Health check (liveness & readiness) probes using an HTTPGetAction will no longer follow redirects to different hostnames from the original probe request. Instead, these non-local redirects will be treated as a Success (the documented behavior). In this case an event with reason "ProbeWarning" will be generated, indicating that the redirect was ignored. If you were previously relying on the redirect to run health checks against different endpoints, you will need to perform the healthcheck logic outside the Kubelet, for instance by proxying the external endpoint rather than redirecting to it.
```

/assign @thockin @liggitt 
/cc @redbaron 
/sig node network
/priority important-soon

### CODE REVIEW

File: test/utils/image/manifest.go [hunk 1]
```diff
@@ -215,7 +215,7 @@ func initImageConfigs() map[int]Config {
 	configs[Iperf] = Config{e2eRegistry, "iperf", "1.0"}
 	configs[JessieDnsutils] = Config{e2eRegistry, "jessie-dnsutils", "1.0"}
 	configs[Kitten] = Config{e2eRegistry, "kitten", "1.0"}
-	configs[Liveness] = Config{e2eRegistry, "liveness", "1.0"}
+	configs[Liveness] = Config{e2eRegistry, "liveness", "1.1"}
```
Comment (thockin) [inline test/utils/image/manifest.go]: Has this been pushed already?

**Review by thockin — COMMENTED** [review]
It's an ugly compromise, but I guess that's what a compromise is.

/lgtm
/approve

File: test/utils/image/manifest.go (hunk 1 above)
Comment (tallclair) [reply test/utils/image/manifest.go]: Yeah.

**Comment by spiffxp** [comment]
/milestone v1.14
v1.14 release lead here, this is the sort of change I wouldn't want to see cherry-picked into v1.14.1 as it changes behavior... and this seems worthwhile to include now instead of waiting until v1.15

File: pkg/probe/http/http.go [hunk 2]
```diff
@@ -102,9 +113,30 @@ func DoHTTPProbe(url *url.URL, headers http.Header, client GetHTTPInterface) (pr
...
 	return probe.Failure, fmt.Sprintf("HTTP probe failed with statuscode: %d", res.StatusCode), nil
 }
+
+func redirectChecker(followNonLocalRedirects bool) func(*http.Request, []*http.Request) error {
+	if followNonLocalRedirects {
+		return nil // Use the default http client checker.
+	}
+
+	return func(req *http.Request, via []*http.Request) error {
+		if req.URL.Hostname() != via[0].URL.Hostname() {
```
[the remark this answers is not in the document]
Comment (tallclair) [reply pkg/probe/http/http.go]: Yes, the URL in the request is parsed relative to the previous request: https://golang.org/src/net/http/client.go?s=18538:18870#L563

This case is also verified by the unit test.

**Pull request merged**